core/src/main/java/org/apache/brooklyn/core/config/BasicConfigInheritance.java

```diff
@@ -170,12 +324,103 @@ public String getConflictResolutionStrategy() {
         return conflictResolutionStrategy;
     }
     
+    @Deprecated /** @deprecated since 0.10.0 when it was introduced, prefer {@link #getLocalDefaultResolvesWithAncestorValue()} */
     public boolean getUseLocalDefaultValue() {
-        return useLocalDefaultValue;
+        return getLocalDefaultResolvesWithAncestorValue();
+    }
+
+    /** see {@link #localDefaultResolvesWithAncestorValue} */
+    public boolean getLocalDefaultResolvesWithAncestorValue() {
+        if (localDefaultResolvesWithAncestorValue==null) {
+            // in case some legacy path is using an improperly deserialized object
+            log.warn("Encountered legacy "+this+" with null localDefaultResolvesWithAncestorValue; transforming", new Throwable("stack trace for legacy "+this));
+            readResolve();
+        }
+        return localDefaultResolvesWithAncestorValue;
+    }
+    
+    public boolean getAncestorDefaultInheritable() {
+        if (ancestorDefaultInheritable==null) {
+            log.warn("Encountered legacy "+this+" with null ancestorDefaultInheritable; transforming", new Throwable("stack trace for legacy "+this));
+            readResolve();
+        }
+        return ancestorDefaultInheritable;
+    }
+
+    // standard deserialization method
+    private ConfigInheritance readResolve() {
+        try {
+            if (useLocalDefaultValue!=null) {
+                // move away from useLocalDefaultValue to localDefaultResolvesWithAncestorValue
+                
+                Field fNew = getClass().getDeclaredField("localDefaultResolvesWithAncestorValue");
+                fNew.setAccessible(true);
+                Field fOld = getClass().getDeclaredField("useLocalDefaultValue");
+                fOld.setAccessible(true);
+                
+                if (fNew.get(this)==null) {
+                    fNew.set(this, useLocalDefaultValue);
+                } else {
+                    if (!fNew.get(this).equals(useLocalDefaultValue)) {
+                        throw new IllegalStateException("Incompatible values detected for "+fOld+" ("+fOld.get(this)+") and "+fNew+" ("+fNew.get(this)+")");
+                    }
+                }
+                fOld.set(this, null);
+            }
+            
+            if (ancestorDefaultInheritable==null) {
+                Field f = getClass().getDeclaredField("ancestorDefaultInheritable");
+                f.setAccessible(true);
+                f.set(this, true);
+            }
+            
+        } catch (Exception e) {
+            throw Exceptions.propagate(e);
+        }
+                
+        return returnEquivalentConstant(this);
     }
 
     @Override
+    public boolean equals(Object obj) {
+        if (obj==null) return false;
+        if (obj instanceof DelegatingConfigInheritance) return equals( ((DelegatingConfigInheritance)obj).getDelegate() );
+        if (obj.getClass().equals(BasicConfigInheritance.class)) {
```

Comment:
This seems wrong, why not `instanceof`?
You already had to work around it in the delegate.